### DISCUSSION TITLE
CefSharp based Blazor hybrid

### DISCUSSION THREAD

**Discussion by jpmikkers** [cefsharp/CefSharp#4440] [Show and tell]
Hi,

I managed to create a CefSharp based WinForms component that enables building Blazor hybrid apps using CefSharp as the underlying webview. I'm wondering if people here see a use case for this?

For those interested, the code is here  [https://github.com/jpmikkers/Baksteen.Blazor.CefSharp](https://github.com/jpmikkers/Baksteen.Blazor.CefSharp)

![image](https://user-images.githubusercontent.com/10578746/228517405-c6f1d63f-ad04-43b1-8b23-82f4fb81beaa.png)

Update: moved from ~~https://github.com/jpmikkers/Baksteen.Avalonia.Blazor/blob/wip/Baksteen.Blazor.CefSharpWinForms/BlazorCefWebView.cs~~

**Comment by amaitland** [comment]
> I managed to create a CefSharp based WinForms component that enables building Blazor hybrid apps using CefSharp as the underlying webview

Nice one! Thanks for sharing. Always interesting to see how people are using `CefSharp` 👍 

> https://github.com/jpmikkers/Baksteen.Avalonia.Blazor/blob/wip/Baksteen.Blazor.CefSharpWinForms/CefCoreWebView2Adapter.cs#L68

As a passing comment if you use an instance of [EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/EventTarget#examples) you'd get built in add/remove/dispatch functionality.

**Reply by jpmikkers** [reply]
@amaitland thanks for the tip, works great and it's much cleaner.

**Comment by joriverm** [comment]
Hey @jpmikkers ,

me and my dev team are currently struggling with this, as the known Airspace issue with WPF is also there for blazorwebview (as it is webview2 based...).
since we are using libcef to do our webview stuff we were wondering if we could use that libcef based webview outside of Avalonia, as this isn't what we are working with (WPF).

i think a CefSharp implementation of the blazor webview would be really beneficial tbh

**Reply by jpmikkers** [reply]
My implementation doesn't rely on Avalonia, it's a winforms component. I just placed it in that repo because it was one of my experiments to see if I could build a cross-platform Blazor webview. I must admit I didn't check my CefSharp based component for airspace issues. If it does, CefSharp's windowlessrendering mode could be a solution, it just renders everything to a bitmap that you could paint into your application.

**Reply by joriverm** [reply]
In general, libcef/cefsharp's ChromiumWebBrowser control does not have the airspace issues so i would assume that if the webview is replaced with cefsharp it *should* work. ill have to clone your repo and test it out then :)

from the looks of it, it looks like the code from Microsoft, but more abstract so you can plug in libcef, correct?

**Reply by jpmikkers** [reply]
Correct! 👍

**Reply by joriverm** [reply]
Hi!

Ive been trying to get a WPF/CefSharp Poc going, as this might solve all our issues.
however, im just getting a white screen, and no idea why.

do you (@jpmikkers ) or @amaitland see an issue?
ive included a patch/git diff of your repo 
[WPF.patch](https://github.com/cefsharp/CefSharp/files/14101082/WPF.patch)

Also, if this works, im willing to create a package/project/whatever to distribute if required

**Reply by jpmikkers** [reply]
@joriverm After applying your patch the CefSharpWinFormsBlazorApp project still works. Your patch file didn't include your WPF projects so I can't see what the issue is. I do remember getting blank screens a lot when getting my [WPF/Blazor template](https://github.com/jpmikkers/Baksteen.Templates.WPF.Blazor) (WPF Blazor hybrid project) up and running, it's quite a finicky process.

**Reply by joriverm** [reply]
ah whoops, forgot to add the files before doing a git diff lol

try this : 
[WPF.patch](https://github.com/cefsharp/CefSharp/files/14111672/WPF.patch)

i have it loading the index html but now the blazor app isn't loading in yet -_-

**Reply by jpmikkers** [reply]
got it running ~~(except for some css issue)~~ , but I tried about a million things so now I have to reconstruct what the main issue was :)

For starters:
- in index.html you need to refer to DemoAppWpf.styles.css
- in index.html autostart must be set to false or it will try to execute the blazor javascript before the callbacks to c# are established
- BSxDispatcher CheckAccess() can't just return false, I used  `public override bool CheckAccess() => _dispatcher.Thread == Thread.CurrentThread;`

**Reply by joriverm** [reply]
if it helps, i noticed in my debug output it had issues dispatching and handling a message from the blazor.webview.js : 
```
requestUri here is https://0.0.0.0/_framework/blazor.modules.json
HandleWebResourceRequest(uri=https://0.0.0.0/_framework/blazor.modules.json)
main frame finished loading
'DemoAppWpf.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.0\System.Diagnostics.StackTrace.dll'. 
'DemoAppWpf.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.0\System.Security.Cryptography.X509Certificates.dll'. 
'DemoAppWpf.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.0\System.Security.Cryptography.dll'. 
webview_consolemessage(Error Uncaught (in promise) TypeError: window.external.receiveMessage is not a function)
'DemoAppWpf.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.0\System.IO.MemoryMappedFiles.dll'. 
```

**Reply by jpmikkers** [reply]
@joriverm I pushed my fixes to the [wip branch](https://github.com/jpmikkers/Baksteen.Avalonia.Blazor/tree/wip) . It seems there's something in the AppBuilder that Blazor needs, so I added a program.cs to do that. Let me know if it works for you too.

**Reply by joriverm** [reply]
hi @jpmikkers !
i can indeed confirm it works now and i feel silly to only realise now the autostart might have been the issue >_<

i created a patch that cleans up some code of the wip branch & moved the website/blazor stuff to a seperate project so its shared between all front ends.
As you can see in the WPF demo, the airspace issue is not there as it is using the CEF WPF control! thats awesome!
[WPF.patch](https://github.com/cefsharp/CefSharp/files/14126137/WPF.patch)

i think it could be *very* beneficial if we made this into a package for others to use 
something like : 
* CefSharp.Webview.Abstraction (Baksteen.Blazor)
* CefSharp.Webview.Wpf (Baksteen.Blazor.CefSharpWPF)
* CefSharp.Webview.Winforms (Baksteen.Blazor.CefSharpWinForms)
* CefSharp.Webview.Avalonia (Baksteen.Avalonia.Blazor)

does this sound good to you @amaitland ?

those packages could then be used by applications to use the BlazorWebview :)

i will be using this stuff for my client, but imo it would be awesome if this was a public thing :)

**Reply by joriverm** [reply]
(whoops, forgot to include my patch)

**Reply by jpmikkers** [reply]
how about I just create a new repo with this stuff (without the Avalonia bits) for visibility and inevitable bugreports :)
See: https://github.com/jpmikkers/Baksteen.Blazor.CefSharp

**Reply by amaitland** [reply]
> i think it could be _very_ beneficial if we made this into a package for others to use
> something like :

If you are suggesting `Nuget package` names then you'd need to include `Blazor` in the name to clearly communicate their intended purpose. As an example `CefSharp.Webview.Abstraction` doesn't tell me what it's for. The name is way too generic, you'll only create confusion.

Reminds me I need to finish the process of reserving the `CefSharp` prefix on `Nuget.org`.

> those packages could then be used by applications to use the BlazorWebview :)
> i will be using this stuff for my client, but imo it would be awesome if this was a public thing :)

I'd start with a public repo first and see what sort of community interest you get. Live with the project for a while, ideally use it in multiple projects so you have a better sense of different use cases. Then review/refine your public `API`.

Publishing/maintaining `Nuget` packages can be a lot of work, make sure you are prepared to the time commitment.

**Reply by joriverm** [reply]
@jpmikkers : i created a solution based on your WIP, and that uses CefSharp as a basis for the webview. it all works, except for the avalonia demo, which is showing a black screen.
from what i gather, avalonia doesn't like the CefSharp winforms control, as i can reproduce it by using a ChromiumBrowser in the avalonia webview instead of a normal webview2 from microsoft. got an idea?

here is the zip : 
[CefSharp.BlazorWebview.zip](https://github.com/cefsharp/CefSharp/files/14168470/CefSharp.BlazorWebview.zip)

as to what you said @amaitland : 
i agree that managing a nuget package takes time and effort. I manage a few nuget packages in my free time and they do take time.
I also agree to what you said in the name being too abstract. I changed this to `CefSharp.BlazorWebview.*` instead of `CefSharp.Webview.*`
the `abstraction` is inline with other packages, like in aspcore, that have a base package that is pure for implementations. things like `Microsoft.Extensions.DependencyInjection.Abstractions` or something :)

so far this looks like a perfect copy of the MS' BlazorWebView and i don't think it would need much changes except for some package changes, but yes, we could maybe first get a repo going and then see :)

**Reply by jpmikkers** [reply]
The new repo is here:
https://github.com/jpmikkers/Baksteen.Blazor.CefSharp

I got rid of the avalonia project, because my original repo covers that.

**Reply by joriverm** [reply]
@jpmikkers : i just noticed a bug in the code. it seems the blazor website can not be interacted with. or thats what it looks like, but in the console log it looks like it does work and clicking links (even if page isn't rendered). seems the UI bit is broken.
got an idea? im currently looking at the dispatcher...

**Reply by jpmikkers** [reply]
@joriverm I've opened a discussions on the repo https://github.com/jpmikkers/Baksteen.Blazor.CefSharp/discussions/1 , lets continue there.
